bcda/encryption/encryption.go

```diff
@@ -0,0 +1,99 @@
+package encryption
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"github.com/CMSgov/bcda-app/bcda/database"
+	"github.com/CMSgov/bcda-app/bcda/models"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"io/ioutil"
+	"os"
+)
+
+// Code in this file borrows heavily from https://github.com/gtank/cryptopasta
+
+// NewEncryptionKey generates a random 256-bit key for Encrypt() and
+// Decrypt(). It panics if the source of randomness fails.
+func newEncryptionKey() *[32]byte {
+	key := [32]byte{}
+	_, err := io.ReadFull(rand.Reader, key[:])
+	if err != nil {
+		panic(err)
+	}
+	return &key
+}
+
+// Encrypt encrypts data using 256-bit AES-GCM.  This both hides the content of
+// the data and provides a check that it hasn't been altered. Output takes the
+// form nonce|ciphertext|tag where '|' indicates concatenation.
+func encrypt(plaintext []byte, key *[32]byte) (ciphertext []byte, err error) {
+	block, err := aes.NewCipher(key[:])
+	if err != nil {
+		return nil, err
+	}
+
+	gcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := make([]byte, gcm.NonceSize())
+	_, err = io.ReadFull(rand.Reader, nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	return gcm.Seal(nonce, nonce, plaintext, nil), nil
+}
+
+func EncryptBytes(publicKey *rsa.PublicKey, plaintext []byte, label string) (ciphertext []byte, encryptedKey []byte, err error) {
+	symmetricKey := newEncryptionKey()
+	ciphertext, err = encrypt(plaintext, symmetricKey)
+	if err != nil {
+		return nil, nil, err
+	}
+	encryptedKey, err = rsa.EncryptOAEP(sha256.New(), rand.Reader, publicKey, symmetricKey[:], []byte(label))
+
+	if err != nil {
+		return nil, nil, err
+	} else {
+		return ciphertext, encryptedKey, nil
+	}
+}
+
+func EncryptAndMove(fromPath, toPath, fileName string, key *rsa.PublicKey, jobID uint) error {
+	// Open and read the file
+	fileBytes, err := ioutil.ReadFile(fromPath + "/" + fileName)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	// Encrypt the file and get the encrypted key
+	encryptedFile, encryptedKey, err := EncryptBytes(key, fileBytes, fileName)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	// Save the encrypted key before trying anything dangerous
+	db := database.GetGORMDbConnection()
+	err = db.Create(&models.JobKey{JobID: jobID, EncryptedKey: encryptedKey, FileName: fileName}).Error
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	// Write out the file to the file system
+	err = ioutil.WriteFile(toPath+"/"+fileName, encryptedFile, os.ModePerm)
```

Comment:
I see this writing the encrypted file to the new location but is it removing the file from the previous location as well?